src/libcore/pin.rs

```diff
@@ -94,6 +95,106 @@
+//! If your type relies on pinning (for example, because it contains internal
+//! references, or because you are implementing something like the intrusive
+//! doubly-linked list mentioned in the previous section), you have to be careful
+//! when implementing `Drop`: notice that `drop` takes `&mut self`, but this
+//! will be called even if your type was previously pinned! It is as if the
+//! compiler automatically called `get_unchecked_mut`. This can never cause
+//! a problem in safe code because implementing a type that relies on pinning
+//! requires unsafe code, but be aware that deciding to make use of pinning
+//! in your type (for example by implementing some operation on `Pin<&[mut] Self>`)
+//! has consequences for your `Drop` implemenetation as well.
+//!
+//! # Projections and Structural Pinning
+//!
+//! One interesting question arises when considering pinning and "container types" --
+//! types such as `Vec` or `Box` but also `RefCell`; types that serve as wrappers
+//! around other types.  When can such a type have a "projection" operation, an
+//! operation with type `fn(Pin<&[mut] Container<T>>) -> Pin<&[mut] T>`?
+//! This does not just apply to generic container types, even for normal structs
+//! the question arises whether `fn(Pin<&[mut] Struct>) -> Pin<&[mut] Field>`
+//! is an operation that can be soundly added to the API.
+//!
+//! This question is closely related to the question of whether pinning is "structural":
+//! when you have pinned a container, have you pinned its contents? Adding a
+//! projection to the API answers that question with a "yes" by offering pinned access
+//! to the contents.
+//!
+//! In general, as the author of a type you get to decide whether pinning is structural, and
+//! whether projections are provided. However, there are a couple requirements to be
+//! upheld when adding projection operations:
+//!
+//! 1. The container must only be [`Unpin`] if all its fields are `Unpin`. This is the default,
+//!    but `Unpin` is a safe trait, so as the author of the container it is your responsibility
+//!    *not* to add something like `impl<T> Unpin for Container<T>`. (Notice that adding a
+//!    projection operation requires unsafe code, so the fact that `Unpin` is a safe trait
+//!    does not break the principle that you only have to worry about any of this if
+//!    you use `unsafe`.)
+//! 2. The destructor of the container must not move out of its argument. This is the exact
+//!    point that was raised in the [previous section][drop-impl]: `drop` takes `&mut self`,
+//!    but the container (and hence its fields) might have been pinned before.
+//!    You have to guarantee that you do not move a field inside your `Drop` implementation.
+//! 3. Your container type must *not* be `#[repr(packed)]`. Packed structs have their fields
+//!    moved around when they are dropped to properly align them, which is in conflict with
+//!    claiming that the fields are pinned when your struct is.
```

Comment:
> Packed structs have their fields moved around when they are dropped to properly align them

Whoa, somehow I've never heard this before.

But shouldn't this be up in the "`Drop` implementation" section? It applies to any type with `#[repr(packed)]` and a `Drop` impl, not just containers with projections, right?

Reply:
Good point.